src/lib/marshal/email-layout/writer.ts

```diff
@@ -0,0 +1,187 @@
+import path from "node:path";
+
+import * as fs from "fs-extra";
+import { get, set, uniqueId } from "lodash";
+
+import { sandboxDir } from "@/lib/helpers/const";
+import { DirContext } from "@/lib/helpers/fs";
+import { DOUBLE_SPACES } from "@/lib/helpers/json";
+import { ObjKeyOrArrayIdx, omitDeep } from "@/lib/helpers/object";
+import { AnyObj, split } from "@/lib/helpers/object";
+import { EmailLayoutDirContext } from "@/lib/run-context";
+
+import { ExtractionSettings, WithAnnotation } from "../shared/types";
+import { EmailLayoutData } from "./types";
+
+export type EmailLayoutDirBundle = {
+  [relpath: string]: string;
+};
+
+type CompiledExtractionSettings = Map<ObjKeyOrArrayIdx[], ExtractionSettings>;
+
+/* Traverse a given email layout data and compile extraction settings of every extractable
+   field into a sorted map.
+*/
+const compileExtractionSettings = (
+  emailLayout: EmailLayoutData<WithAnnotation>,
+): CompiledExtractionSettings => {
+  const extractableFields = get(
+    emailLayout,
+    ["__annotation", "extractable_fields"],
+    {},
+  );
+  const map: CompiledExtractionSettings = new Map();
+
+  for (const [key] of Object.entries(emailLayout)) {
+    // If the field we are on is extractable, then add its extraction
+    // settings to the map with the current object path.
+    if (key in extractableFields) {
+      map.set([key], extractableFields[key]);
+    }
+  }
+
+  return map;
+};
+
+/* Sanitize the email latyout content into a format that's appropriate for reading
+   and writing, by stripping out any annotation fields and handling readonly
+   fields.
+*/
+const toEmailLayoutJson = (
+  emailLayout: EmailLayoutData<WithAnnotation>,
+): AnyObj => {
+  // Move read only field under the dedicated field "__readonly".
+  const readonlyFields = emailLayout.__annotation?.readonly_fields || [];
+  const [readonly, remainder] = split(emailLayout, readonlyFields);
+  const emailLayoutjson = { ...remainder, __readonly: readonly };
+
+  // Strip out all schema annotations, so not to expose them to end users.
+  return omitDeep(emailLayoutjson, ["__annotation"]);
+};
+
+/* Builds an email layout dir bundle, which consist of the email layout JSON + the extractable files.
+   Then writes them into a layout directory on a local file system.
+*/
+export const writeEmailLayoutDirFromData = async (
```

Comment:
I used the approach we took on the `workflows pull` command, but adapted to the email layouts.
In this case is simpler, so the idea is to build a `bundle` which consist of:

* The `html_layout` html extracted file.
*  The `text_layout` txt extracted file.
* The `email_layout_key.json` JSON file.

One we have all that we insert the given files in the indicated `emailLayoutDirCtx  path`